File: app/src/main/java/za/co/lukestonehm/logicaldefence/MainActivity.java
package za.co.lukestonehm.logicaldefence;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.res.Configuration;
import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;


public class MainActivity extends AppCompatActivity
        implements NavCallback {


    public static final String TAG = "LogicalDefence";

    private CharSequence mTitle;
    private String[] sections;

    private int mCurrentSelectedPosition = 0;

    private DrawerLayout mDrawerLayout;
    private ActionBarDrawerToggle mDrawerToggle;
    private ListView mDrawerListView;

    String[] locales;
    String[] languages;
    NavCallback mCallbacks;
    Menu currentActionViewMenu;

    AppPreferences appPrefs;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        drawScreen();
    }

    private void drawScreen() {
        setContentView(R.layout.activity_main);
        mCallbacks = this;
        appPrefs = new AppPreferences(this);

        mDrawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
        mDrawerLayout.setDrawerShadow(R.drawable.drawer_shadow, GravityCompat.START);
        locales = getResources().getStringArray(R.array.supported_locales);
        languages = getResources().getStringArray(R.array.locale_names);

        setLocale(appPrefs.getLocale());

        //load sections after locale setup
        sections = getResources().getStringArray(R.array.sections);

        Toolbar mActionBar = (Toolbar) findViewById(R.id.action_bar);
        setSupportActionBar(mActionBar);

        mDrawerToggle = new ActionBarDrawerToggle(
                this,                    /* host Activity */
                mDrawerLayout,                    /* DrawerLayout object */
                R.string.navigation_drawer_open,  /* "open drawer" description for accessibility */
                R.string.navigation_drawer_close  /* "close drawer" description for accessibility */
        ) {
            @Override
            public void onDrawerClosed(View drawerView) {
                super.onDrawerClosed(drawerView);
            }

            @Override
            public void onDrawerOpened(View drawerView) {
                setDrawerItemSelected((TextView) mDrawerListView.getChildAt(mCurrentSelectedPosition));
                super.onDrawerOpened(drawerView);
            }
        };

        mDrawerLayout.setDrawerListener(mDrawerToggle);
        ActionBar ab = getSupportActionBar();
        if (ab != null)
            ab.setDisplayHomeAsUpEnabled(true);

        mDrawerToggle.syncState();

        mDrawerListView = (ListView) findViewById(R.id.lv);
        mDrawerListView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                unsetDrawerItemSelected((TextView) parent.getChildAt(mCurrentSelectedPosition));
                setDrawerItemSelected((TextView) view);
                selectItem(position);
            }
        });

        mDrawerListView.setAdapter(new ArrayAdapter<>(
                this,
                R.layout.nav_item,
                R.id.nav_item, sections
        ));

        mCurrentSelectedPosition = appPrefs.getSection();

        mTitle = sections[mCurrentSelectedPosition];
        setTitle(mTitle);
        mDrawerListView.setItemChecked(mCurrentSelectedPosition, true);
        changeSection(mCurrentSelectedPosition);
    }


    private void selectItem(int position) {
        mCurrentSelectedPosition = position;

        mDrawerListView.setItemChecked(position, true);
        mDrawerLayout.closeDrawers();
        mCallbacks.onNavigationDrawerItemSelected(position);

        if (appPrefs == null)
            appPrefs = new AppPreferences(this);

        appPrefs.setSection(position);

        mTitle = sections[position];
        setTitle(mTitle);
    }

    private void changeSection(int position) {
        FragmentManager fragmentManager = getSupportFragmentManager();
        fragmentManager.beginTransaction()
                .replace(R.id.container, PlaceholderFragment.newInstance(position + 1))
                .commit();
    }

    private void setDrawerItemSelected(TextView item) {
        item.setBackgroundColor(getResources().getColor(R.color.colorPrimary));
        item.setTextColor(Color.WHITE);
    }

    private void unsetDrawerItemSelected(TextView item) {
        item.setBackgroundColor(Color.TRANSPARENT);
        item.setTextColor(getResources().getColor(R.color.primary_dark_material_dark));
    }

    @Override
    public void onNavigationDrawerItemSelected(int position) {
        // update the main content by replacing fragments
        changeSection(position);
    }

    public void onSectionAttached(int number) {
        if (sections == null)
            sections = getResources().getStringArray(R.array.sections);
        mTitle = sections[number - 1];
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        currentActionViewMenu = menu;
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        if (mDrawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        int id = item.getItemId();

        if (id == R.id.action_about) {
            startActivity(new Intent(this, About.class));
            return true;
        }

        if (id == R.id.action_language) {

            //Create Alert dialog to choose language from
            AlertDialog.Builder builder = new AlertDialog.Builder(this);
            builder.setTitle(R.string.language_dialog_title);
            builder.setItems(languages, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialog, int which) {
                    setLocale(which);
                    Intent refresh = new Intent(getApplicationContext(), MainActivity.class);
                    startActivity(refresh);
                    finish();
                }
            });
            builder.show();

            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    /**
     * Set the selected language
     *
     * @param which position in locale array
     * @see <a href="http://stackoverflow.com/questions/12908289/how-to-change-language-of-app-when-user-selects-language">how-to-change-language-of-app-when-user-selects-language</a>
     */
    public void setLocale(int which) {
        if (which > -1) {
            Locale newLocale = new Locale(locales[which]);
            Configuration config = getResources().getConfiguration();
            config.locale = newLocale;
            getResources().updateConfiguration(config, getResources().getDisplayMetrics());

            appPrefs.setLocale(which);//save locale
        }
    }

    /**
     * A placeholder fragment containing a simple view.
     */
    public static class PlaceholderFragment extends Fragment {
        /**
         * The fragment argument representing the section number for this
         * fragment.
         */
        private static final String ARG_SECTION_NUMBER = "section_number";

        /**
         * Returns a new instance of this fragment for the given section
         * number.
         */
        public static PlaceholderFragment newInstance(int sectionNumber) {
            PlaceholderFragment fragment = new PlaceholderFragment();
            Bundle args = new Bundle();
            args.putInt(ARG_SECTION_NUMBER, sectionNumber);
            fragment.setArguments(args);
            return fragment;
        }

        public PlaceholderFragment() {
        }

        @Override
        public View onCreateView(LayoutInflater inflater, ViewGroup container,
                                 Bundle savedInstanceState) {
            View v = inflater.inflate(R.layout.fragment_main, container, false);

            String[] titles = getStringArrByName("fallacies_titles_" + getArguments().getInt(ARG_SECTION_NUMBER));
            String[] descs = getStringArrByName("fallacies_descs_" + getArguments().getInt(ARG_SECTION_NUMBER));
            String[] examples = getStringArrByName("fallacies_examples_" + getArguments().getInt(ARG_SECTION_NUMBER));

            List<Fallacy> fallacies = generateFallacyList(titles, descs, examples);

            RecyclerView mRecyclerView = (RecyclerView) v.findViewById(R.id.sec_rv);
            mRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
            FallacyListAdapter fla = new FallacyListAdapter(getActivity(), fallacies);
            mRecyclerView.setAdapter(fla);

            return v;
        }

        private String[] getStringArrByName(String name) {
            int i = getResources().getIdentifier(name, "array",
                    getActivity().getApplicationContext().getPackageName()); // You had used "name"
            return getResources().getStringArray(i);
        }

        @Override
        public void onAttach(Activity activity) {
            super.onAttach(activity);
            ((MainActivity) activity).onSectionAttached(
                    getArguments().getInt(ARG_SECTION_NUMBER));
        }

        private List<Fallacy> generateFallacyList(String[] titles, String[] descs, String[] examples) {
            List<Fallacy> list = new ArrayList<>();
            for (int i = 0; i < titles.length; i++) {
                list.add(new Fallacy(i, titles[i], descs[i], examples[i]));
            }
            return list;
        }
    }

}

File: app/src/main/java/za/co/lukestonehm/logicaldefence/MainActivity.kt
package za.co.lukestonehm.logicaldefence

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.*
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import java.util.*

class MainActivity : AppCompatActivity(), NavCallback {
    private var mTitle: CharSequence? = null
    private var sections: Array<String>? = null
    private var mCurrentSelectedPosition = 0
    private var mDrawerLayout: DrawerLayout? = null
    private var mDrawerToggle: ActionBarDrawerToggle? = null
    private var mDrawerListView: ListView? = null
    var locales: Array<String> = arrayOf()
    var languages: Array<String> = arrayOf()
    var mCallbacks: NavCallback? = null
    var currentActionViewMenu: Menu? = null
    var appPrefs: AppPreferences? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        drawScreen()
    }

    private fun drawScreen() {
        setContentView(R.layout.activity_main)
        mCallbacks = this
        appPrefs = AppPreferences(this)
        mDrawerLayout = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        mDrawerLayout!!.setDrawerShadow(R.drawable.drawer_shadow, GravityCompat.START)
        locales = resources.getStringArray(R.array.supported_locales)
        languages = resources.getStringArray(R.array.locale_names)
        setLocale(appPrefs!!.locale)

        //load sections after locale setup
        sections = resources.getStringArray(R.array.sections)
        val mActionBar = findViewById<Toolbar>(R.id.action_bar)
        setSupportActionBar(mActionBar)
        mDrawerToggle = object : ActionBarDrawerToggle(
                this,  /* host Activity */
                mDrawerLayout,  /* DrawerLayout object */
                R.string.navigation_drawer_open,  /* "open drawer" description for accessibility */
                R.string.navigation_drawer_close /* "close drawer" description for accessibility */
        ) {
            override fun onDrawerClosed(drawerView: View) {
                super.onDrawerClosed(drawerView)
            }

            override fun onDrawerOpened(drawerView: View) {
                setDrawerItemSelected(mDrawerListView!!.getChildAt(mCurrentSelectedPosition) as TextView)
                super.onDrawerOpened(drawerView)
            }
        }
        mDrawerLayout!!.setDrawerListener(mDrawerToggle)
        val ab = supportActionBar
        ab?.setDisplayHomeAsUpEnabled(true)
        mDrawerToggle!!.syncState()
        mDrawerListView = findViewById<View>(R.id.lv) as ListView
        mDrawerListView!!.onItemClickListener = OnItemClickListener { parent, view, position, id ->
            unsetDrawerItemSelected(parent.getChildAt(mCurrentSelectedPosition) as TextView)
            setDrawerItemSelected(view as TextView)
            selectItem(position)
        }
        mDrawerListView!!.adapter = ArrayAdapter(
                this,
                R.layout.nav_item,
                R.id.nav_item, sections!!
        )
        mCurrentSelectedPosition = appPrefs!!.section
        mTitle = sections!![mCurrentSelectedPosition]
        title = mTitle
        mDrawerListView!!.setItemChecked(mCurrentSelectedPosition, true)
        changeSection(mCurrentSelectedPosition)
    }

    private fun selectItem(position: Int) {
        mCurrentSelectedPosition = position
        mDrawerListView!!.setItemChecked(position, true)
        mDrawerLayout!!.closeDrawers()
        mCallbacks!!.onNavigationDrawerItemSelected(position)
        if (appPrefs == null) appPrefs = AppPreferences(this)
        appPrefs!!.section = position
        mTitle = sections!![position]
        title = mTitle
    }

    private fun changeSection(position: Int) {
        val fragmentManager = supportFragmentManager
        fragmentManager.beginTransaction()
                .replace(R.id.container, PlaceholderFragment.newInstance(position + 1))
                .commit()
    }

    private fun setDrawerItemSelected(item: TextView?) {
        if (item != null) {
            item.setBackgroundColor(resources.getColor(R.color.colorPrimary))
            item.setTextColor(Color.WHITE)
        }
    }

    private fun unsetDrawerItemSelected(item: TextView?) {
        if (item != null) {
            item.setBackgroundColor(Color.TRANSPARENT)
            item.setTextColor(resources.getColor(R.color.primary_dark_material_dark))
        }
    }

    override fun onNavigationDrawerItemSelected(position: Int) {
        // update the main content by replacing fragments
        changeSection(position)
    }

    fun onSectionAttached(number: Int) {
        if (sections == null) sections = resources.getStringArray(R.array.sections)
        mTitle = sections!![number - 1]
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        currentActionViewMenu = menu
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        if (mDrawerToggle!!.onOptionsItemSelected(item)) {
            return true
        }
        val id = item.itemId
        if (id == R.id.action_about) {
            startActivity(Intent(this, About::class.java))
            return true
        }
        if (id == R.id.action_language) {

            //Create Alert dialog to choose language from
            val builder = AlertDialog.Builder(this)
            builder.setTitle(R.string.language_dialog_title)
            builder.setItems(languages) { dialog, which ->
                setLocale(which)
                val refresh = Intent(applicationContext, MainActivity::class.java)
                startActivity(refresh)
                finish()
            }
            builder.show()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * Set the selected language
     *
     * @param which position in locale array
     * @see [how-to-change-language-of-app-when-user-selects-language](http://stackoverflow.com/questions/12908289/how-to-change-language-of-app-when-user-selects-language)
     */
    fun setLocale(which: Int) {
        if (which > -1) {
            val newLocale = Locale(locales[which])
            val config = resources.configuration
            config.locale = newLocale
            resources.updateConfiguration(config, resources.displayMetrics)
            appPrefs!!.locale = which //save locale
        }
    }

    /**
     * A placeholder fragment containing a simple view.
     */
    class PlaceholderFragment : Fragment() {
        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                                  savedInstanceState: Bundle?): View? {
            val v = inflater.inflate(R.layout.fragment_main, container, false)
            val titles = getStringArrByName("fallacies_titles_" + arguments!!.getInt(ARG_SECTION_NUMBER))
            val descs = getStringArrByName("fallacies_descs_" + arguments!!.getInt(ARG_SECTION_NUMBER))
            val examples = getStringArrByName("fallacies_examples_" + arguments!!.getInt(ARG_SECTION_NUMBER))
            val fallacies = generateFallacyList(titles, descs, examples)
            val mRecyclerView = v.findViewById<View>(R.id.sec_rv) as RecyclerView
            mRecyclerView.layoutManager = LinearLayoutManager(activity)
            val fla = FallacyListAdapter(activity!!, fallacies)
            mRecyclerView.adapter = fla
            return v
        }

        private fun getStringArrByName(name: String): Array<String> {
            val i = resources.getIdentifier(name, "array",
                    activity!!.applicationContext.packageName) // You had used "name"
            return resources.getStringArray(i)
        }

        override fun onAttach(activity: Activity) {
            super.onAttach(activity)
            (activity as MainActivity).onSectionAttached(
                    arguments!!.getInt(ARG_SECTION_NUMBER))
        }

        private fun generateFallacyList(titles: Array<String>, descs: Array<String>, examples: Array<String>): List<Fallacy> {
            val list: MutableList<Fallacy> = ArrayList()
            for (i in titles.indices) {
                list.add(Fallacy(i, titles[i], descs[i], examples[i]))
            }
            return list
        }

        companion object {
            /**
             * The fragment argument representing the section number for this
             * fragment.
             */
            private const val ARG_SECTION_NUMBER = "section_number"

            /**
             * Returns a new instance of this fragment for the given section
             * number.
             */
            fun newInstance(sectionNumber: Int): PlaceholderFragment {
                val fragment = PlaceholderFragment()
                val args = Bundle()
                args.putInt(ARG_SECTION_NUMBER, sectionNumber)
                fragment.arguments = args
                return fragment
            }
        }
    }

    companion object {
        const val TAG = "LogicalDefence"
    }
}